File: ruby/easy/108_convert_sorted_array_to_bst.rb
# 108. Convert Sorted Array to Binary Search Tree
# Easy
#
# Given an integer array nums where the elements are sorted in ascending
# order, convert it to a height-balanced binary search tree.
#
# A height-balanced binary tree is a binary tree in which the depth of the
# two subtrees of every node never differs by more than one.

def sorted_array_to_bst(nums)
  if nums.nil? || nums.size.zero?
    nil
  elsif nums.size == 1
    TreeNode.new(nums.pop, nil, nil)
  else
    m = nums.size / 2
    TreeNode.new(nums[m], sorted_array_to_bst(nums[0...m]), sorted_array_to_bst(nums[m + 1..]))
  end
end

File: go/108_convert_sorted_array_to_bst.go
// 108. Convert Sorted Array to Binary Search Tree
// Easy
//
// Given an integer array nums where the elements are sorted in ascending
// order, convert it to a height-balanced binary search tree.
//
// A height-balanced binary tree is a binary tree in which the depth of the
// two subtrees of every node never differs by more than one.
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sortedArrayToBST(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}

	mid := len(nums) / 2

	return &TreeNode{
		Val:   nums[mid],
		Left:  sortedArrayToBST(nums[:mid]),
		Right: sortedArrayToBST(nums[mid+1:]),
	}
}
